mbed_host_tests/host_tests/base_host_test.py

```diff
@@ -15,28 +15,116 @@
+from time import time
+
+
+class BaseHostTestAbstract(object):
     """ Base class for each host-test test cases with standard
         setup, test and teardown set of functions
     """
 
     name = ''   # name of the host test (used for local registration)
+    event_queue = None      # To main even loop
+    dut_event_queue = None  # To DUT
 
-    def setup(self):
-        """ Set up function, initialize your test case dynamic resources in this function
+    def _notify_prn(self, text, nl=True):
+        if self.event_queue:
+            if nl:
+                text += '\n'
+            self.event_queue.put(('__notify_prn', text, time()))
+
+    def notify_complete(self, consume=True):
+        """! Notify htrun that host test finished processing
+        @param consume If True htrun will process (consume) all remaining events
         """
+        if self.event_queue:
+            self.event_queue.put(('__notify_complete', consume, time()))
+
+    def log(self, text):
+        self._notify_prn(text)
+
+    def _notify_dut(self, key, value):
+        """! Send data over serial to DUT """
+        if self.event_queue:
+            self.dut_event_queue.put((key, value, time()))
+
+    def send_kv(self, key, value):
+        self._notify_dut(key, value)
+
+    def setup(self):
+        pass
+
+    def test(self):
         pass
 
-    def test(self, selftest):
-        """ Blocking test execution process:
+    def teardown(self):
+        pass
+
+
+class HostTestCallbackBase(BaseHostTestAbstract):
+
+    def __init__(self):
+        BaseHostTestAbstract.__init__(self)
+        self.callbacks = {}
+        self._restricted_callbacks = ['coverage_start',
+            'testcase_start',
+            'testcase_finish'
+            ]
+
+        self.printable = ['coverage_start',
+            'testcase_start',
+            'testcase_finish'
+            ]
+
+        self._assign_callbacks()
+
+    def _callback_default(self, key, value, timestamp):
+        """! Default callback """
+        self.log("CALLBACK: key=%s, value=%s, timestamp=%f"% (key, value, timestamp))
+
+    def _callback_forward(self, key, value, timestamp):
+        """! We want to print on stdout things Greentea can capture"""
+        # TODO:
+        pass
 
-            setup()
-            test()
-            teardown()
+    def _assign_callbacks(self):
+        """! Assigns default callback handlers
         """
+        for key in self.printable:
+            self.callbacks[key] = self._callback_forward
+
+    def register_callback(self, key, callback):
+        if type(key) is not str:
+            raise TypeError
+
+        if key in self._restricted_callbacks:
+            raise ValueError
+
+        if not callable(callback):
+            raise TypeError
+
+        self.callbacks[key] = callback
+
+    def setup(self):
+        pass
+
+    def test(self):
+        pass
+
+    def teardown(self):
+        pass
+
+
+class BaseHostTest(HostTestCallbackBase):
```

Comment:
which of these 3 classes in this file are intended to be used by a user? all 3? why setup,test,teardown are empty in all these classes?